zerver/decorator.py

```diff
@@ -681,32 +698,30 @@ def wrapper(func):
         def wrapped_func(request, *args, **kwargs):
             # type: (HttpRequest, *Any, **Any) -> HttpResponse
 
-            # It is really tempting to not even wrap our original function
-            # when settings.RATE_LIMITING is False, but it would make
-            # for awkward unit testing in some situations.
-            if not settings.RATE_LIMITING:
-                return func(request, *args, **kwargs)
-
             if client_is_exempt_from_rate_limiting(request):
                 return func(request, *args, **kwargs)
 
-            try:
-                user = request.user
-            except Exception:
-                # TODO: This logic is not tested, and I'm not sure we are
-                # doing the right thing here.
-                user = None
-
-            if not user:
-                logging.error("Requested rate-limiting on %s but user is not authenticated!" %
-                              func.__name__)
-                return func(request, *args, **kwargs)
-
-            # Rate-limiting data is stored in redis
-            # We also only support rate-limiting authenticated
-            # views right now.
-            # TODO(leo) - implement per-IP non-authed rate limiting
-            rate_limit_user(request, user, domain)
+            if settings.RATE_LIMITING:
+                # User based rate limiting. Works when a user is logged in.
+                try:
+                    user = request.user
+                except Exception:
+                    # Not authed. Ignore and move to IP based rate limiting.
+                    logging.debug("Requested rate-limiting on {} but user is "
+                                  "not authenticated!".format(func.__name__))
+                else:
+                    # This function can raise RateLimited exception which is
+                    # processed by
+                    # zerver.middleware.RateLimitMiddleware.process_exception.
+                    rate_limit_user(request, user, domain)
+
+            if settings.IP_RATE_LIMITING:
+                ip = get_ip(request)
```

Comment:
Yeah, I tracked down the repo and took a quick look. The fact that there is essentially zero documentation of what it is that this code is supposed to actually accomplish (other than in words like "best" or "real") does not make me super confident in relying on it.

Also they're using `str.startswith` on string representations of fragments of IP addresses to attempt prefix-matching, which is asking for trouble -- and for example, this code as it is will mistake an address like `::1:0123:4567:89ab:cdef:0123:4567` (aka `0000:0001:0123:4567:89ab:cdef:0123:4567`) for a loopback address because its abbreviated form bears a passing resemblance to that of `::1` (aka `0000:0000:0000:0000:0000:0000:0000:0001`).

So I'm now leaning more toward the view that IP rate-limiting should be either disabled in the default config, or configured to do something very simple; and admins who want to use it at scale should expect to play around with it a bit to find the right config.